storage, server: don't reuse leases obtained before a restart or transfer

andreimatei opened the pull request:
Before this patch, leases held before a restart could be used after the
restart. This is incorrect, since the command queue has been wiped in
the restart and so reads and writes are not properly sequenced with
possible in-flight commands.

There was also a problem before this patch, related to lease transfers:
if a transfer command returned an error to the `replica.Send()` call
that proposed it, we'd clear our "in transfer" state and allow the
replica to use the existing lease. However, such errors are generally
ambiguous - there's no guarantee that the transfer will still apply. In
such cases, the replica was actually breaking its promise to not use the
lease after initiating the transfer.

The current patch addresses both these problems by introducing a
`replica.minLeaseProposedTs` and a `Lease.ProposedTs`. These fields are
used to enforce the fact that only leases proposed after a particular
time will be used *for proposing* new commands, or for serving reads. On
a transfer, the field for the replica in question is set to the present
time. On restart, the field is set for all the replicas to the current
time, after the Server has waited such that it has a guarantee that the
current HLC is above what it might have been on previous incarnations.
This ensures that, in both the transfer and the restart case, leases
that have been proposed before the transfer/restart and were in flight
at the time when the transfer/restart happens are not eligible to be
used.

This patch also changes the way waiting on startup for a server's HLC to
be guaranteed to be monotonic wrt to the HLC before the restart works.
Before, we were waiting just before starting serving, but after all the
stores had been started. The point of the waiting was to not serve reads
thinking we have a lower timestamp than before the restart. I'm unclear
about whether that was correct or not, considering that, for example, a
store's queues were already running and potentially doing work.
Now that we also rely on that waiting for not reusing old leases (we
need to initialize replica.minLeaseProposedTs with a value higher than
the node had before the restart), I've had to do the waiting *before*
initializing the stores. I could have done the setting of that field
late, but it seems even more dangerous than before to allow queues to do
work with potentially bad leases.
We lost the amortization of this wait time with the store creation
process... But note that, for empty engines (e.g. in tests) we don't do
any waiting.

Fixes #7996

THIS PATCH NEEDS TO BE DEPLOYED THROUGH A "STOP THE WORLD"
Because of the field being added to the Lease proto, we can't have
leases proposed by new servers being applied by old servers - they're
going to be serialized differently and the consistency checker will flag
it.
Note that a "freeze" is not required, though. The new field is nullable,
so any lease requests that might be in flight at the time of the world
restart will be serialized the same as they have been by the nodes that
already applied them before the restart.

cc @petermattis @tschottdorf @bdarnell 
@tamird for validating what I've said above about not needing the freeze
@jseldess for keeping in mind that the beta where we roll this out with need a "stop the world"

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10420)
<!-- Reviewable:end -->

andreimatei committed:
storage, server: don't reuse leases obtained before a restart or transfer

Before this patch, leases held before a restart could be used after the
restart. This is incorrect, since the command queue has been wiped in
the restart and so reads and writes are not properly sequenced with
possible in-flight commands.

There was also a problem before this patch, related to lease transfers:
if a transfer command returned an error to the `replica.Send()` call
that proposed it, we'd clear our "in transfer" state and allow the
replica to use the existing lease. However, such errors are generally
ambiguous - there's no guarantee that the transfer will still apply. In
such cases, the replica was actually breaking its promise to not use the
lease after initiating the tranfer.

The current patch addresses both these problems by introducing a
`replica.minLeaseProposedTS` and a `Lease.ProposedTS`. These fields are
used to enforce the fact that only leases proposed after a particular
time will be used *for proposing* new commands, or for serving reads. On
a transfer, the field for the replica in question is set to the present
time. On restart, the field is set for all the replicas to the current
time, after the Server has waited such that it has a guarantee that the
current HLC is above what it might have been on previous incarnations.
This ensures that, in both the transfer and the restart case, leases
that have been proposed before the transfer/restart and were in flight
at the time when the transfer/restart happens are not eligible to be
used.

This patch also changes the way waiting on startup for a server's HLC to
be guaranteed to be monotonic wrt to the HLC before the restart works.
Before, we were waiting just before starting serving, but after all the
stores had been started. The point of the waiting was to not serve reads
thinking we have a lower timestamp than before the restart. I'm unclear
about whether that was correct or not, considering that, for example, a
store's queues were already running and potentially doing work.
Now that we also rely on that waiting for not reusing old leases (we
need to initialize replica.minLeaseProposedTS with a value higher than
the node had before the restart), I've had to do the waiting *before*
initializing the stores. I could have done the setting of that field
late, but it seems even more dangerous than before to allow queues to do
work with potentially bad leases.
We lost the amortization of this wait time with the store creation
process... But note that, for empty engines (e.g. in tests) we don't do
any waiting.

Fixes #7996

THIS PATCH NEEDS TO BE DEPLOYED THROUGH A "STOP THE WORLD"
Because of the field being added to the Lease proto, we can't have
leases proposed by new servers being applied by old servers - they're
going to be serialized differently and the consistency checker will flag
it.
Note that a "freeze" is not required, though. The new field is nullable,
so any lease requests that might be in flight at the time of the world
restart will be serialized the same as they have been by the nodes that
already applied them before the restart.

andreimatei committed:
storage: factor out common test function

andreimatei commented:
Review status: 0 of 13 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/storage/replica.go, line 3647 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVb2aQdxLmDfmdxAcRY:-KVb2aQeCg6Ewjz_ivvB:b-2j903w) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica.go#L3647)):*
> ```Go
> // particular, we shouldn't be gossiping if we previously initiated a transfer
> // and then died.
> func (r *Replica) leaseIsValidForGossiping() bool {
> ```

I'm not entirely clear on when exactly we want to gossip and when we don't want to... I'm not entirely sure of the consequences of gossiping at the wrong time. Maybe @tschottdorf or someone else can suggest a better comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420)*
<!-- Sent from Reviewable.io -->

tamird commented:
Needs a rebase and has some failing tests. Should we be reviewing before the tests are passing?

andreimatei commented:
Yeah, please review. The tests are about the proto that changed, and
something about range stats differing from a golden number, I think.

Btw I'm also writing a test for a server restart, but that needs some new
test infrastructure.

On Nov 2, 2016 10:05 PM, "Tamir Duberstein" <notifications@github.com>
wrote:

> Needs a rebase and has some failing tests. Should we be reviewing before
> the tests are passing?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/10420#issuecomment-258050334>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXBcZSELg5IJ-w1xPJvRJbRfTUxIeTvks5q6UF3gaJpZM4Knytg>
> .
>

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 13 of 13 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/roachpb/data.proto, line 335 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdPKsX4tfpHUdyqt73:-KVdPKsX4tfpHUdyqt74:b73carh) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/roachpb/data.proto#L335)):*
> ```protobuf
>   optional ReplicaDescriptor replica = 3 [(gogoproto.nullable) = false];
> 
>   // The current timestamp of the proposed at the time when this lease has been
> ```

s/of the proposed//

---

*[pkg/storage/replica.go, line 618 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdUJIR2w5Hok9UP8p5:-KVdUJIR2w5Hok9UP8p6:bu0eb03) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica.go#L618)):*
> ```Go
> 	// overridden.
> 	if !r.store.cfg.TestingKnobs.DontPreventUseOfOldLeaseOnStart {
> 		r.mu.minLeaseProposedTs = clock.Now()
> ```

This will give all the replicas (slightly) different timestamps. This could be a single value chosen by the store, although that's probably not worth the plumbing. 

---

*[pkg/storage/replica.go, line 3647 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVb2aQdxLmDfmdxAcRY:-KVdQuXCFBnMLEyvG_uS:b-3krmxu) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica.go#L3647)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I'm not entirely clear on when exactly we want to gossip and when we don't want to... I'm not entirely sure of the consequences of gossiping at the wrong time. Maybe @tschottdorf or someone else can suggest a better comment.
</details>
Gossip is inherently inconsistent and asynchronous, so while it would be bad for a node to gossip long after its lease has expired, all the subtleties of the HLC are largely unnecessary here. We're just using the lease as a way to ensure that only one node gossips at a time. I think this check is correct - we should stop gossiping after we have transferred the lease away.

We don't really need a gossip-specific definition of lease validity; this could use the `haveLease` method being introduced in #10305. 

---

*[pkg/storage/replica_test.go, line 492 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdX4fphpkcieQDwkFp:-KVdX4fphpkcieQDwkFq:b-z1i55t) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_test.go#L492)):*
> ```Go
> // Test the behavior of a replica while a range lease transfer is in progress:
> // - while the transfer is in progress, reads should return errors pointing to
> // the transfer targer
> ```

s/targer/target/

---

*[pkg/storage/store.go, line 144 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdVzx4KIrrAvsA_M8i:-KVdVzx4KIrrAvsA_M8j:b-gp3ioq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/store.go#L144)):*
> ```Go
> 			// the restart, to make their life easy. They're not concerned with the
> 			// reasons why normally we dissalow those leases from being used.
> 			DontPreventUseOfOldLeaseOnStart: true,
> ```

Can we set this in those tests instead of for all tests that use TestStoreConfig?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-:-KVdXbNt7qKhUo9OQUmR:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> good test (ok, bit hard to follow along, but that's just how those tests are). Does this affect the runtime of the chaos acceptance tests at all? Those are the only tests that would still potentially eat longer unavailability because of `kill -9`, any others would have nodes (at least in the future) transfer all of their leases away before shutdown.

---

Reviewed 13 of 13 files at r2.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[pkg/roachpb/data.proto, line 335 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdPKsX4tfpHUdyqt73:-KVdvgij4maaYcyQzRUI:b-pvtroo) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/roachpb/data.proto#L335)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/of the proposed//
</details>
give this `(gogoproto.customname) = "ProposedTS"` and then `s/Ts/TS/` throughout (including commit message).

---

*[pkg/roachpb/data.proto, line 340 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdvpQfIKaD3fbszahO:-KVdvpQfIKaD3fbszahP:b-dbwpjo) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/roachpb/data.proto#L340)):*
> ```protobuf
>   // This is nullable to help with the rollout (such that a lease applied by
>   // some nodes before the rollout and some nodes after the rollout is
>   // serialized the same). It should be non-nullable subsequently.
> ```

Make this `TODO(andrei)`.

Also, while you're here, you may want to fix the ordering - I think we generally want to order by tag number for simplicity (?). I definitely just wasted time thinking why you'd skipped `4`.

---

*[pkg/server/server.go, line 561 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdwC9sy6PtS6jk7w8o:-KVdwC9sy6PtS6jk7w8p:b-doge5w) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/server/server.go#L561)):*
> ```Go
> 	// information. For example, a client might have written a value at a
> 	// timestamp that's in the future of the restarted node's clock, and if we
> 	// don't do something, the same client's read would not return it. So, we
> ```

`it` is a little too ambiguous, perhaps `s/it/their write/`.

---

*[pkg/server/server.go, line 562 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdwInWzCPK2o44bsZM:-KVdwInWzCPK2o44bsZN:b-hoh3ci) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/server/server.go#L562)):*
> ```Go
> 	// timestamp that's in the future of the restarted node's clock, and if we
> 	// don't do something, the same client's read would not return it. So, we
> 	// wait up to MaxOffset. We assume we couldn't have served timestamps more
> ```

We don't assume, we know.

---

*[pkg/storage/replica.go, line 297 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdwTDzz9aW-2efYdq9:-KVdwTDzz9aW-2efYdqA:b-2w286i) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica.go#L297)):*
> ```Go
> 		// minLeaseProposedTs is the minimum acceptable proposedTs lease; only
> 		// leases proposed after this timestamp can be used for proposing commands.
> 		minLeaseProposedTs hlc.Timestamp
> ```

`s/Ts/TS/g` including the commit message.

---

*[pkg/storage/replica.go, line 3647 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVb2aQdxLmDfmdxAcRY:-KVdy009KE-9Uo8cZNXS:b99cpnn) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica.go#L3647)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Gossip is inherently inconsistent and asynchronous, so while it would be bad for a node to gossip long after its lease has expired, all the subtleties of the HLC are largely unnecessary here. We're just using the lease as a way to ensure that only one node gossips at a time. I think this check is correct - we should stop gossiping after we have transferred the lease away.
> 
> We don't really need a gossip-specific definition of lease validity; this could use the `haveLease` method being introduced in #10305. 
</details>
Agreed, though I think there's some value in making it more transparent what we can use the lease for (`haveLease` is pretty ambiguous again). Maybe something like `canPropose` is more descriptive. Well, a bike shed for another day. Removing the copy pasta in `maybeGossip.*` is definitely a step up.

---

*[pkg/storage/replica_range_lease.go, line 139 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdySERh3lX-PYMaHP2:-KVdySERh3lX-PYMaHP3:bgtn6d2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_range_lease.go#L139)):*
> ```Go
> 		// transfer will not still apply. That's OK, however, as the "in transfer"
> 		// state maintained by the pendingLeaseRequest is not relied on for
> 		// correctness, and reseting the state is beneficial as it'll allow the
> ```

s/reset/resett/

Maybe hint at the mechanism that does ensure correctness.

---

*[pkg/storage/replica_range_lease.go, line 193 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdyd4NwxorWCJVKH_y:-KVdyd4NwxorWCJVKH_z:bo462be) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_range_lease.go#L193)):*
> ```Go
> // correctness: if a previous transfer has returned an error, TransferInProgress
> // will return `false`, but that doesn't necessarily mean that the transfer
> // cannot still apply.
> ```

Maybe hint at the mechanism that does ensure correctness.

---

*[pkg/storage/replica_test.go, line 298 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdythiH2ePToQ75-Ng:-KVdythiH2ePToQ75-Nh:bhuixh9) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_test.go#L298)):*
> ```Go
> 	rngDesc := *tc.rng.Desc()
> 	rngDesc.Replicas = append(rngDesc.Replicas, secondReplica)
> 	tc.rng.setDescWithoutProcessUpdate(&rngDesc)
> ```

I know this isn't a new hack (and thanks for the de-pasta'ing), but no bueno: this lets in-memory and on-disk state diverge and is something someone might have to clean up and shake their head in disbelief over in the future.
I think you can still do this, but run a put that persists this descriptor as well (then call `r.assertState()` at the end to see whether it checks out).

---

*[pkg/storage/replica_test.go, line 501 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVe-XGe70Ozdeh59aWf:-KVe-XGe70Ozdeh59aWg:b-11khug) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_test.go#L501)):*
> ```Go
> 	tc := testContext{}
> 	tsc := TestStoreConfig()
> 	var leaseAcquisitionTrap func(ts hlc.Timestamp)
> ```

nit: might looked better as an `atomic.Value`

---

*[pkg/storage/replica_test.go, line 508 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVe-g1_a7ryf7rSBVm6:-KVe-g1_a7ryf7rSBVm7:ba0o5f3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_test.go#L508)):*
> ```Go
> 			leaseAcquisitionTrap(ts)
> 		}
> 		defer leaseAcquisitionTrapMu.Unlock()
> ```

could go away with `atomic.Value`, but if not move it adjacent to the `Lock()`.

---

*[pkg/storage/replica_test.go, line 511 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVe-qecJUlyzj0E7pQp:-KVe-qecJUlyzj0E7pQq:b-e0q674) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_test.go#L511)):*
> ```Go
> 	}
> 	transferSem := make(chan struct{})
> 	tsc.TestingKnobs.TestingCommandFilter =
> ```

This one is *probably* fine with proposer-evaluated KV (since you're doing single-node stuff and don't really care *where* in the proposal you trap the thing).

---

*[pkg/storage/replica_test.go, line 519 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVe0KFWkTlKBGWfyjIo:-KVe0KFWkTlKBGWfyjIp:b-s96zo0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_test.go#L519)):*
> ```Go
> 				<-transferSem
> 				// Return an error, so that the pendingLeaseRequest considers the
> 				// transferred failed.
> ```

s/red//

---

*[pkg/storage/replica_test.go, line 553 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVe2kVlqwUaydWos0lP:-KVe2kVlqwUaydWos0lQ:b-9b8179) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_test.go#L553)):*
> ```Go
> 	repDesc, err := tc.rng.getReplicaDescriptorLocked()
> 	if err != nil {
> 		t.Fatal(err)
> ```

missing Unlock.

---

*[pkg/storage/replica_test.go, line 592 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVe32O423C3G6ElaY5w:-KVe32O423C3G6ElaY5x:ba5pts0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_test.go#L592)):*
> ```Go
> 	// apply.
> 	// Concretely, we're going to check that a read triggers a new lease
> 	// acqusition.
> ```

`acquisition`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-:-KVe470-267ZAtkJWHBy:b-z19dcc)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r1, 10 of 13 files at r2.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[pkg/server/server.go, line 562 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdwInWzCPK2o44bsZM:-KVfmGUuuMgzylaQ0aB1:bfw4x3o) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/server/server.go#L562)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> We don't assume, we know.
</details>
No we don't - the previous incarnation could have been running with a higher MaxOffset.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
still haven't written that restart test... but I will.
And haven't checked the chaos tests, will do.

---

Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[pkg/roachpb/data.proto, line 335 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdPKsX4tfpHUdyqt73:-KVespncF9MGY59taDyt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/roachpb/data.proto#L335)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> give this `(gogoproto.customname) = "ProposedTS"` and then `s/Ts/TS/` throughout (including commit message).
</details>
Done.

---

*[pkg/roachpb/data.proto, line 340 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdvpQfIKaD3fbszahO:-KVeqStZvEOmuWKtw0A2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/roachpb/data.proto#L340)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Make this `TODO(andrei)`.
> 
> Also, while you're here, you may want to fix the ordering - I think we generally want to order by tag number for simplicity (?). I definitely just wasted time thinking why you'd skipped `4`.
</details>
Done.

---

*[pkg/server/server.go, line 561 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdwC9sy6PtS6jk7w8o:-KVet4bbbrr33nIuFIu-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/server/server.go#L561)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> `it` is a little too ambiguous, perhaps `s/it/their write/`.
</details>
Done.

---

*[pkg/server/server.go, line 562 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdwInWzCPK2o44bsZM:-KVetN93ndqY5urzuxvk:bm8ba2w) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/server/server.go#L562)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> We don't assume, we know.
</details>
Done.
"JERRY: Well, why would I assume. I never assume. Leads to assumptions."

---

*[pkg/storage/replica.go, line 297 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdwTDzz9aW-2efYdq9:-KVetTxidIxJ7NPo30MI:bjhiid6) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica.go#L297)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> `s/Ts/TS/g` including the commit message.
</details>
Done.
But... why tho? "Timestamp" is one word...

---

*[pkg/storage/replica.go, line 618 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdUJIR2w5Hok9UP8p5:-KVetaUW9x03kbQwxKEd:b-phwxxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica.go#L618)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This will give all the replicas (slightly) different timestamps. This could be a single value chosen by the store, although that's probably not worth the plumbing. 
</details>
yeah, I've considered that. In fact that's how I wrote it initially, but, yeah, too much plumbing.
Also these replicas already use divergent timestamps for stuff (e.g. the tsCache low-water mark)

---

*[pkg/storage/replica.go, line 3647 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVb2aQdxLmDfmdxAcRY:-KVexMgLJkj4Sydgs__1:b-97q0s3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica.go#L3647)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Agreed, though I think there's some value in making it more transparent what we can use the lease for (`haveLease` is pretty ambiguous again). Maybe something like `canPropose` is more descriptive. Well, a bike shed for another day. Removing the copy pasta in `maybeGossip.*` is definitely a step up.
</details>
I've renamed the function and rewritten the comment; see if you like it.
I don't want to introduce more consolidation of checks before #10305...

---

*[pkg/storage/replica_range_lease.go, line 139 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdySERh3lX-PYMaHP2:-KVey9zN3NZADSOsVC6Q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_range_lease.go#L139)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/reset/resett/
> 
> Maybe hint at the mechanism that does ensure correctness.
</details>
Done.

---

*[pkg/storage/replica_range_lease.go, line 193 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdyd4NwxorWCJVKH_y:-KVeyHUkjtjweLsJ9fze:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_range_lease.go#L193)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Maybe hint at the mechanism that does ensure correctness.
</details>
Done.

---

*[pkg/storage/replica_test.go, line 298 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdythiH2ePToQ75-Ng:-KVg-zkoahmt6TIN-BXg:b-a4urld) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_test.go#L298)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I know this isn't a new hack (and thanks for the de-pasta'ing), but no bueno: this lets in-memory and on-disk state diverge and is something someone might have to clean up and shake their head in disbelief over in the future.
> I think you can still do this, but run a put that persists this descriptor as well (then call `r.assertState()` at the end to see whether it checks out).
</details>
Done.
Although shaking your head in disbelief is definitely part of the job :)

---

*[pkg/storage/replica_test.go, line 492 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdX4fphpkcieQDwkFp:-KVeyo9XPGb2-1AIgKi0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_test.go#L492)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/targer/target/
</details>
Done.

---

*[pkg/storage/replica_test.go, line 501 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVe-XGe70Ozdeh59aWf:-KVg0v3-bRqorMcyY3mW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_test.go#L501)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> nit: might looked better as an `atomic.Value`
</details>
Done.

---

*[pkg/storage/replica_test.go, line 508 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVe-g1_a7ryf7rSBVm6:-KVg0vWLIe37dcmfodcJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_test.go#L508)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> could go away with `atomic.Value`, but if not move it adjacent to the `Lock()`.
</details>
Done.

---

*[pkg/storage/replica_test.go, line 511 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVe-qecJUlyzj0E7pQp:-KVeyvCYkV-ifiUObS5z:b-5ctcay) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_test.go#L511)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> This one is *probably* fine with proposer-evaluated KV (since you're doing single-node stuff and don't really care *where* in the proposal you trap the thing).
</details>
right... But this comment sounds like it's in contrast with something else not being fine with prop-eval KV... But I don't see that something.

---

*[pkg/storage/replica_test.go, line 519 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVe0KFWkTlKBGWfyjIo:-KVez4rQS3XPoU6xuaWO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_test.go#L519)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/red//
</details>
Done.

---

*[pkg/storage/replica_test.go, line 553 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVe2kVlqwUaydWos0lP:-KVezJUd_efPnCQdnbpt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_test.go#L553)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> missing Unlock.
</details>
Done.

---

*[pkg/storage/replica_test.go, line 592 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVe32O423C3G6ElaY5w:-KVezOEOPJYQwpk8usFp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_test.go#L592)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> `acquisition`.
</details>
Done.

---

*[pkg/storage/store.go, line 144 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdVzx4KIrrAvsA_M8i:-KVg7OYzaGbfS_D91LHQ:blw95hv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/store.go#L144)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Can we set this in those tests instead of for all tests that use TestStoreConfig?
</details>
Well, the problem is that no test fails without this :) So I was afraid I was changing test behavior and invalidating tests.
But I've looked through the tests doing restarts, and only one seemed to want to preserve leases. The others were in `client_raft_test.go` and `client_split_test.go`, and, at least on the very surface, don't seem like they'd care about the state of leases. So I've let them be.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-:-KVg9gq7aqinfuWfwJgE:b-1pelzn)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 14 of 14 files at r3, 12 of 15 files at r4.
Review status: 13 of 16 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[pkg/server/server.go, line 562 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdwInWzCPK2o44bsZM:-KVgDoqriUahft0XCBeF:b-enxk80) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/server/server.go#L562)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Done.
> "JERRY: Well, why would I assume. I never assume. Leads to assumptions."
</details>
again, the previous incarnation could have had a higher maxoffset. you should also link to that issue you filed about this.

---

*[pkg/storage/client_raft_test.go, line 309 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVgE0HL-WSf8ljYi04W:-KVgE0HL-WSf8ljYi04X:bq8afn8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b2a30a5842d38773267517946793e4d54ce01bc/pkg/storage/client_raft_test.go#L309)):*
> ```Go
> 	// range can cause spurious lease transfers which cause this test to fail.
> 	sc.TestingKnobs.DisablePeriodicGossips = true
> 	sc.TestingKnobs.DontPreventUseOfOldLeaseOnStart = true
> ```

comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 14 files at r3, 15 of 15 files at r4.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/server/server.go, line 562 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdwInWzCPK2o44bsZM:-KVgFn0tWjwh7yCn4XQZ:bfnq6rx) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/server/server.go#L562)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> again, the previous incarnation could have had a higher maxoffset. you should also link to that issue you filed about this.
</details>
Yeah, a comment about what would need to be done to support changing of `MaxOffset` safely would be good, though the main comment of that type maybe already exists elsewhere. Perhaps just link the issue as @tamird suggested.

---

*[pkg/storage/replica.go, line 297 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdwTDzz9aW-2efYdq9:-KVgG3bE6imQu0a7mMiY:b-jttwvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica.go#L297)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Done.
> But... why tho? "Timestamp" is one word...
</details>
Hypertext is also a word and yet it is `HTTP`, not `HtTP`.

---

*[pkg/storage/replica_test.go, line 511 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVe-qecJUlyzj0E7pQp:-KVgGampdKmeEJDwGiEl:bqgfheg) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/storage/replica_test.go#L511)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> right... But this comment sounds like it's in contrast with something else not being fine with prop-eval KV... But I don't see that something.
</details>
Sorry, just an isolated comment. Basically "I looked at this because I didn't want it to break with proposer-evaluated KV, and I don't think it would".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 17 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/server/server.go, line 562 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVdwInWzCPK2o44bsZM:-KVgconq3tHQCyk3Kwjp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6cca1f80029098175099842634a0ab3839b24852/pkg/server/server.go#L562)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Yeah, a comment about what would need to be done to support changing of `MaxOffset` safely would be good, though the main comment of that type maybe already exists elsewhere. Perhaps just link the issue as @tamird suggested.
</details>
Done.

---

*[pkg/storage/client_raft_test.go, line 309 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVgE0HL-WSf8ljYi04W:-KVgczPnO_N6o34hDgvu:b6anhis) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b2a30a5842d38773267517946793e4d54ce01bc/pkg/storage/client_raft_test.go#L309)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> comment.
</details>
well the comment from the definition of the field applies sufficiently well

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 16 files at r5, 12 of 16 files at r6.
Review status: 13 of 17 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/storage/client_raft_test.go, line 309 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVgE0HL-WSf8ljYi04W:-KVhI1l-TuhAjE-DPLRx:b-rg6syp) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b2a30a5842d38773267517946793e4d54ce01bc/pkg/storage/client_raft_test.go#L309)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> well the comment from the definition of the field applies sufficiently well
</details>
it doesn't explain what this test does that requires this flag

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
the only chaos test that's not skipped is TestNodeRestart, which takes anywhere between 15 and 22s, with or without this PR.

---

Review status: 0 of 17 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/storage/client_raft_test.go, line 309 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVgE0HL-WSf8ljYi04W:-KVzhhm9DUniQdFvY2jv:bvil3t9) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b2a30a5842d38773267517946793e4d54ce01bc/pkg/storage/client_raft_test.go#L309)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> it doesn't explain what this test does that requires this flag
</details>
added a comment, but I think it's as redundant as they get.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-:-KVzhZUlbFEFkZv6aQ2e:b-d0fwi0)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 17 files at r7, 13 of 16 files at r8.
Review status: 14 of 17 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_range_lease.go, line 217 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420#-KVzqXGIkyrBEsiMj4-o:-KVzqXGIkyrBEsiMj4-p:be8yd3n) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d508a110e5232fe20e6fa01b0283184211e35b8/pkg/storage/replica_range_lease.go#L217)):*
> ```Go
> // sent on the returned chan.
> func (r *Replica) requestLeaseLocked(timestamp hlc.Timestamp) <-chan *roachpb.Error {
> 	if r.store.TestingKnobs().LeaseRequestEvent != nil {
> ```

slight preference for a binding local to this `if`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10420)*
<!-- Sent from Reviewable.io -->
